internal/security/secretstore/vmkencryption.go

```diff
@@ -0,0 +1,280 @@
+// in the end, Keys and KeysBase64 are removed and replaced with
+// EncryptedKeys and Nonces in the resulting JSON
+// Root token is left untouched
+func (v *VMKEncryption) EncryptInitResponse(initResp *secretstoreclient.InitResponse) error {
+
+	// This is a no-op unless LoadIKM() was called previously
+	if !v.encrypting {
+		return nil
+	}
+
+	newKeys := make([]string, len(initResp.Keys))
+	newNonces := make([]string, len(initResp.Keys))
+
+	for i, hexPlaintext := range initResp.Keys {
+
+		plainText, err := hex.DecodeString(hexPlaintext)
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to decode hex bytes of keyshare (details omitted)"))
+			return err
+		}
+
+		keyShare, nonce, err := v.gcmEncryptKeyshare(plainText, i) // Wrap using a unique AES key
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to wrap key %d: %s", i, err.Error()))
+			return err
+		}
+
+		newKeys[i] = hex.EncodeToString(keyShare)
+		newNonces[i] = hex.EncodeToString(nonce)
+
+		wipeKey(keyShare) // Clear out binary version of encrypted key
+		wipeKey(nonce)    // Clear out nonce
+	}
+
+	initResp.EncryptedKeys = newKeys
+	initResp.Nonces = newNonces
+	initResp.Keys = nil       // strings are immutable, must wait for GC
+	initResp.KeysBase64 = nil // strings are immutable, must wait for GC
+	return nil
+}
+
+// DecryptInitResponse processes the InitResponse and decrypts the key shares
+// in the end, EncryptedKeys and Nonces are removed and replaced with
+// Keys and KeysBase64 in the resulting JSON like the init response was originally
+// Root token is left untouched
+func (v *VMKEncryption) DecryptInitResponse(initResp *secretstoreclient.InitResponse) error {
+
+	// This is a no-op unless LoadIKM() was called previously
+	if !v.encrypting {
+		return nil
+	}
+
+	newKeys := make([]string, len(initResp.EncryptedKeys))
+	newKeysBase64 := make([]string, len(initResp.EncryptedKeys))
+
+	for i, hexCiphertext := range initResp.EncryptedKeys {
+
+		hexNonce := initResp.Nonces[i]
+		nonce, err := hex.DecodeString(hexNonce)
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to decode hex bytes of nonce %s", err.Error()))
+			return err
+		}
+
+		cipherText, err := hex.DecodeString(hexCiphertext)
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to decode hex bytes of ciphertext"))
+			return err
+		}
+
+		keyShare, err := v.gcmDecryptKeyshare(cipherText, nonce, i) // Unwrap using a unique AES key
+		if err != nil {
+			v.logger.Error(fmt.Sprintf("failed to unwrap key %d: %s", i, err.Error()))
+			return err
+		}
+
+		newKeys[i] = hex.EncodeToString(keyShare)
+		newKeysBase64[i] = base64.StdEncoding.EncodeToString(keyShare)
+	}
+
+	initResp.Keys = newKeys
+	initResp.KeysBase64 = newKeysBase64
+	initResp.EncryptedKeys = nil
+	initResp.Nonces = nil
+	return nil
+}
+
+//
+// Internal methods
+//
+
+func (v *VMKEncryption) gcmEncryptKeyshare(keyshare []byte, counter int) ([]byte, []byte, error) {
```

Comment:
this function could use some more comments if only to help verify the pattern implemented here

Reply:
added